src/Features/CSharp/Portable/InternalUtilities/CSharpOperationAnalysisHelpers.cs

```diff
@@ -0,0 +1,941 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.Features
+{
+    internal static class CSharpOperationAnalysisHelpers
+    {
+        public static bool IsSimpleEqualsMethod(
+            Compilation compilation,
+            IMethodSymbol methodSymbol,
+            IMethodBodyOperation methodBodyOperation,
+            ImmutableArray<IFieldSymbol> expectedComparedFields)
+        {
+            if (methodSymbol.Name == nameof(Equals) &&
+                methodSymbol.ReturnType.SpecialType == SpecialType.System_Boolean &&
+                methodSymbol.Parameters.IsSingle())
+            {
+                var type = methodSymbol.ContainingType;
+                var equatableType = GetIEquatableType(compilation, type);
+                if (OverridesEquals(compilation, methodSymbol, equatableType))
+                {
+                    if (equatableType != null &&
+                        methodSymbol.Parameters.First().Type.SpecialType == SpecialType.System_Object &&
+                        (GetBlockOfMethodBody(methodBodyOperation)) is IBlockOperation
+                        {
+                            Operations: [IReturnOperation
+                            {
+                                ReturnedValue: IInvocationOperation
+                                {
+                                    Instance: IInstanceReferenceOperation,
+                                    Arguments: [IArgumentOperation { Value: IOperation arg }]
+                                }
+                            }]
+                        } && arg.WalkDownConversion() is IParameterReferenceOperation { Parameter: IParameterSymbol param }
+                        && param.Equals(methodSymbol.Parameters.First()))
+                    {
+                        // in this case where we have an Equals(C? other) from IEquatable but the current one
+                        // is Equals(object? other), we accept something of the form:
+                        // return Equals(other as C);
+                        return true;
+                    }
+
+                    // otherwise we check to see which fields are compared (either by themselves or through properties)
+                    var actualFields = GetEqualizedFields(methodBodyOperation, methodSymbol);
+                    return !actualFields.HasDuplicates(SymbolEqualityComparer.Default) &&
+                            actualFields.SetEquals(expectedComparedFields);
+                }
+            }
+
+            return false;
+        }
+
+        public static bool IsSimpleHashCodeMethod(
+            Compilation compilation,
+            IMethodSymbol methodSymbol,
+            IMethodBodyOperation methodOperation,
+            ImmutableArray<IFieldSymbol> expectedHashedFields)
+        {
+            if (methodSymbol.Name == nameof(GetHashCode) &&
+                methodSymbol.Parameters.IsEmpty &&
+                UseSystemHashCode.Analyzer.TryGetAnalyzer(compilation, out var analyzer))
```

Comment:
Hmm, I would have assumed if anything the UseSystemHashCode stuff would call into this common helper, not the other way around here. Should the calling order be reversed?

Reply:
So I think I should:
1. Move all the Equality and constructor code into a helper class within the convert to record folder/namespace
2. Move the UseSystemHashCode analyzer logic here, and both this and the original use system hash code can call into here.